utilities/poap/puppet_agent_install.py

```diff
@@ -0,0 +1,239 @@
+    DNS = ''
+
+DNS = DNS.strip()
+# Optional DOMAIN
+if os.environ.get('DOMAIN'):
+    DOMAIN = os.environ['DOMAIN']
+elif 'DOMAIN' not in globals():
+    DOMAIN = ''
+
+DOMAIN = DOMAIN.strip()
+
+# Optionally set os.environ proxies
+proxies = ['HTTP_PROXY', 'HTTPS_PROXY', 'NO_PROXY']
+for p in proxies:
+    if not os.environ.get(p):
+        if p in globals() and len(p):
+            os.environ[p] = globals()[p]
+
+# Create a logfile
+log_prefix = datetime.now().strftime('/bootflash/puppet_agent_install.%Y%H%M%S')
+log_handle = open('%s.log' % log_prefix, "w+")
+
+# Temporary wget html (deleted by script)
+wget_html = '%s.wget' % log_prefix
+
+###############################################################################
+# METHODS
+###############################################################################
+
+def log_it(text):
+    print text
+    log_handle.write('\n' + text)
+    log_handle.flush()
+    sys.stdout.flush()
+
+def log_it_step(text):
+    text = '\n--------\n [STEP] %s' % text
+    print text
+    log_handle.write('\n' + text)
+    log_handle.flush()
+    sys.stdout.flush()
+
+def log_globals():
+    """Log all global variables"""
+
+    all_vars = 'PUPPET_BINARY PUPPET_CONFIGPRINT RPM_NAME RPM_URI \
+                PUPPET_SERVER PREPEND_CMD DNS VRF DOMAIN \
+                HTTP_PROXY HTTPS_PROXY NO_PROXY'
+    buf = ''
+    for v in all_vars.split():
+        buf += '\n%20s = %s' % (v, globals().get(v))
+    log_it('--------\nGlobal Variables:\n' + buf)
+
+def process_cmd(cmd):
+    """Process native bash or guestshell command"""
+
+    log_it('\ncommand> ' + cmd)
+    args = shlex.split(cmd)
+    p = subprocess.Popen(args, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+    output,error = p.communicate()
+    log_it('\n stdout> ' + output)
+    if error:
+        log_it('\n stderr> ' + error)
+    if p.returncode == 0:
+        return output.rstrip()
+    else:
+        return "FAIL: {0}{1}".format(output, error)
+
+def configure_nameserver():
+    """Create /etc/resolv.conf in the agent environment"""
+
+    if not DNS:
+        return
+    resolv = '/etc/resolv.conf'
+    log_it_step('Create config for %s' % resolv)
+    process_cmd('sudo rm -f ' + resolv)
+    log_it('\ncommand> (Write config to %s)' % resolv)
+    with open(resolv, 'w') as f:
+        f.write(DNS)
+    process_cmd('sudo chmod 666 ' + resolv)
+
+def verify_reachability():
+    """Verify network reachability to puppet master and rpm repo """
+
+    log_it_step('Verify reachability to puppet master')
+    cmd = PREPEND_CMD + ' ping -c 5 ' + PUPPET_SERVER
+    result = process_cmd(cmd)
+    if result.find('FAIL') == 0:
+        log_it('Failed to ping puppet master')
+        exit(0)
+
+    log_it_step('Verify reachability to RPM repo (wget writes to stderr)')
+    cmd = '%s wget %s -O %s' % (PREPEND_CMD, RPM_URI, wget_html)
+    result = process_cmd(cmd)
+    if result.find('FAIL') == 0:
+        log_it("Failed to reach RPM repo")
+        exit(0)
```

Comment:
For exit code failure cases, we should return the exit code of the command that failed.